File: bin/Data/EditorLua/view/StartView.lua
local StartView = {}
StartView.isShow = true
local Translator = require("EditorLua/core/Translate")
StartView.app = nil
StartView.trigged = false

function StartView:RemoveHistoryItem(path)
    --self.app.RemoveHistoryItem();
end

function StartView:RenderHistoryList()
    imgui.TextUnformatted("Open Recent:")
    for k,v in ipairs(self.history.attributes._history) do
        -- body
        --ImGuiCol_Button
        imgui.PushStyleColor(imgui.constant.Col.Button,0x00ffffff)
        --ImGuiStyleVar_ButtonTextAlign
        --PushStyleVar
        --button text align left
        imgui.PushStyleVar_2(imgui.constant.StyleVar.ButtonTextAlign,0,0)
        if imgui.Button(v,300,25) then
            self.app.StartGame(v)
        end
        imgui.PopStyleVar(1)
        --imgui.SetNextWindowPos(320,height,0,0,0)
        imgui.SameLine();
        if imgui.Button("x",25,25) then
            self:RemoveHistoryItem(v)
        end
        
        imgui.PopStyleColor(1)
        
    end
end


function StartView:Update()
    if self.isShow == true then
        self.isShow = imgui.Begin(Translator:translate("StartView"),self.isShow)
        --local undotex = self.assetMgr:getImguiTex("res/img/undo.png")
        --imgui.Image(undotex,32,32)
        imgui.PushStyleVar_2(imgui.constant.StyleVar.ButtonTextAlign,0,0)
        if imgui.Button("Open...",80,25) then
            self.trigged = true;
        end
        imgui.PopStyleVar(1)
        imgui.Spacing()
        imgui.Spacing()
        imgui.Spacing()
        imgui.Separator()
        self:RenderHistoryList()
        if self.trigged then
            local path = self.app.DialogGetPath()
            if path ~= "" then
                self.app.StartGame(path)
            end
        end
        self.trigged = false
        imgui.End()
    end
    
end

function StartView:Show(show)
    self.isShow = show
end

return StartView

File: Source/UrhoEditor/mvc/view/StartView.cpp
#include "StartView.h"
#include "imgui.h"
#include "Urho3D/Core/Context.h"
#include "Urho3D/Core/Main.h"
#include "Urho3D/Engine/Engine.h"
#include "imgui_impl_glfw.h"
#include "GLFW/glfw3.h"
#include "imgui_impl_opengl3.h"

namespace Urho3DEditor 
{

StartView::StartView(StartGameHandler startGame, SelectPathHandler selectPath, SetWorkSpaceHandler setWorkSpace, HistoryMgr& historyMgr)
	: startGame_(startGame)
	, selectPath_(selectPath)
	, setWorkSpace_(setWorkSpace)
	, historyMgr_(historyMgr)
{
}

StartView::~StartView() 
{
}

void StartView::RenderHistoryList() 
{
	auto historys = historyMgr_.get_history_list();
	for(auto item: historys)
	{
		ImGui::Text("");
		ImGui::SameLine();
		ImGui::PushStyleColor(ImGuiCol_Button,0x00ffffff);
		ImGui::PushStyleVar(ImGuiStyleVar_ButtonTextAlign,ImVec2(0,0));
		if(ImGui::Button(item.CString()))
		{
			if (setWorkSpace_)
				setWorkSpace_(item);
			if (startGame_)
				startGame_();
			_isShow = false;
		}
		ImGui::PopStyleVar(1);
		ImGui::SameLine();
		if( ImGui::Button("x"))
		{
			historyMgr_.delete_history(item);
			historyMgr_.save();
		}
		ImGui::PopStyleColor(1);
	}
}

void StartView::Update() 
{
	if(!_isShow)
		return;
	ImGui::BeginChild("StartView");
	ImGui::Spacing();
	ImGui::PushStyleVar(ImGuiStyleVar_ButtonTextAlign,ImVec2(0,0));
	ImGui::Text("");
	ImGui::SameLine();
	if (ImGui::Button("Open..."))
	{
		String path = selectPath_ ? selectPath_() : String();
		if(path != "")
		{
			historyMgr_.add_project(path);
			historyMgr_.save();
			if (setWorkSpace_)
				setWorkSpace_(path);
			if (startGame_)
				startGame_();
			_isShow = false;
		}
		
	}
	ImGui::PopStyleVar(1);
	ImGui::Spacing(), ImGui::Spacing(), ImGui::Spacing();
	ImGui::Separator();
	this->RenderHistoryList();
	
	ImGui::EndChild();
}
void StartView::OpenProject() 
{
}
void StartView::OnOpen() 
{
}
}